### DISCUSSION TITLE
PlatformIO support

### DISCUSSION THREAD

**Discussion by CSC-Sendance** [adafruit/Adafruit_nRF52_Arduino#793] [Show and tell]
Many of you probably know platformio as an advanced alternative to the ArduinoIDE, for those who don't: https://platformio.org/ . [PlatformIO also supports the nrf52 series](https://github.com/platformio/platform-nordicnrf52) but particularly its dependencies, like this "framework" repository, needs quite a few hacks to get running. Here is where platformio's [dependency registry](https://registry.platformio.org/) comes into play. It apparently takes this repository and then adds some compatibility files (mainly [`library.json`](https://docs.platformio.org/en/latest/manifests/library-json/) and `package.json` files) by some hidden scripting (?). The [official "platform package" repository page](https://registry.platformio.org/tools/platformio/framework-arduinoadafruitnrf52) also links to this very repository here, which does not have these modifications.

I have scanned the needed modifications and successfully tested them in [my personal fork here](https://github.com/CSC-Sendance/Adafruit_nRF52_Arduino). 

So, what would be needed?

- A package.json in the root repository, see [example here](https://github.com/adafruit/Adafruit_nRF52_Arduino/commit/941901e6262d13fa5b8b4817798711d87ef53f23). E.g., the [esp32 Arduino core repo also has one ](https://github.com/espressif/arduino-esp32/blob/master/package.json) . Maintenance effort: bump semantic version number on every release

- For all git submodule dependencies (nrfCrypto and TinyUSB), a `library.json` is needed and some additional linker stuff that is pio-specific, so shouldn't have an effect on the ArduinoIDE. Maintenance effort: 0 (?)
- for libraries/Adafruit_nRFCrypto, I already made a PR, see https://github.com/adafruit/Adafruit_nRFCrypto/pull/6
- for libraries/Adafruit_TinyUSB_Arduino, I already made a PR, see https://github.com/adafruit/Adafruit_TinyUSB_Arduino/pull/336

All in all, the effort required to maintain support is probably minimal. Is this something that people/maintainers are interested in? The advantage would be that this nrf52 "framework"/Arduino Core can effectively be forked and used by platformio users, and thereby contributions made back to the framework - this is currently NOT possible since using the framework in platformio requires the changes made by the pio-repository (see, e.g. [this discussion](https://community.platformio.org/t/building-with-non-release-platform-package-framework-results-in-dependency-assertion-error/35439/7)). 

P.S.: To try for yourself if the changes suggested above work, include in your `platformio.ini` `[env]`:

```
platform_packages = 
	framework-arduinoadafruitnrf52 @ https://github.com/CSC-Sendance/Adafruit_nRF52_Arduino
```

The only further modification included in this repo is, that I updated the nrfCrypto and TInyUSB dependencies and made [this small fix to get them working](https://github.com/CSC-Sendance/Adafruit_nRF52_Arduino/commit/65e106acc0674804ef38eced249ac7973d1fd67d)

**Comment by hathach** [comment]
Thank you, I am open to have pio support, it is a nice IDE. However, I don't use it much since it does have some difference with Arduino that does effect the compiling e.g prioritize library include path over core's one. Therefore my main concern is maintaing the pio compatibility, we can't guarantee to do so. People could start to complain when thing is not working. Previously, I got a user yelling and even lecturing me with laws that we have to support him since the library is listed on platformio index as compatible (which is indexed by pio crawler). They probably forgot that MIT license software comes with no warranty or whatsoever.

TLDR, I would be happy to review if you are willing to do the PR. Though in the future when lack of maintenance causing too many support, we could revert this.

**Reply by CSC-Sendance** [reply]
Thanks for your reply. I suppose entitlement is always a problem but should not be attributed to the PIO project itself. However, I 
 completely understand your concerns about maintenance. Potentially, the support could be coined "inofficial" and refrain from providing official examples etc. to not "encourage newbies" too much on platformio, but to have them stick with ArduinoIDE - you know what I mean.

If you want I could also ping [@maxgerhardt](https://github.com/maxgerhardt) who is an avid [pio community moderator](https://community.platformio.org/u/maxgerhardt) that helped me with some of these issues. He could potentially give some tips and tricks based on getting support from other platforms to avoid support cases.

Overall, I would, of course, be willing to do the PR. Due to the dependencies mentioned above it would need to include:
1. an updated libraries/Adafruit_nRFCrypto: https://github.com/adafruit/Adafruit_nRFCrypto/commit/41059f507f494b15b945574e8d64070a8ba0e05c
2. an updated adafruit/Adafruit_TinyUSB_Arduino https://github.com/adafruit/Adafruit_TinyUSB_Arduino/commit/5c650e84e3d88e4abd73fc5485c4f80d2b2bd7f7
3. a fix for the Bluefruit52 "lib" due to one of the more recent updates causing some namespace issues (see [this](https://github.com/CSC-Sendance/Adafruit_nRF52_Arduino/commit/65e106acc0674804ef38eced249ac7973d1fd67d))

Do you see any issues with that? My somewhat rudimentary testing with these changes included, did not cause any issues so far.

**Reply by hathach** [reply]
PIO is currently and probably will be not officially supported (unless somehow I could figured out all of its aspect). Though I still and will leave issue open for a couple of months or so for other community member to fix.

For any suggested code changes, please make an PR, it is easier to review code within a PR

**Reply by CSC-Sendance** [reply]
OK. I will make sure to create a clean one in the next few days.
